[Examples] TPU-based training of a language model using TensorFlow

sayakpaul opened the pull request:
This PR adds an example of performing (masked) language model training using TensorFlow and TPUs. The example is meant to act as a reference for the community on this topic. The following are the main components of the PR: 

* Tokenizer training script (for completeness) 
* TFRecords preparation script (recommended practice when using TPUs)
* Training script
* Evaluation / inference

The purpose of this separation (as opposed to having everything in a single script) is to allow the community to have isolated reference points for performing TPU-based training of our models, which I think is beneficial. 

The artifacts produced during this project can be found here: https://huggingface.co/tf-tpu.

* Tokenizer (trained from scratch): https://huggingface.co/tf-tpu/unigram-tokenizer-wikitext
* Model: https://huggingface.co/tf-tpu/roberta-base-epochs-500-no-wd

Cc: @Rocketknight1 @gante @amyeroberts

sayakpaul committed:
add: tokenizer training script for TF TPU LM training.

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_unigram.py:
```diff
@@ -0,0 +1,117 @@
+    )
+    parser.add_argument(
+        "-vs",
+        "--vocab_size",
```
Maybe we should play around with this a bit to see if a multiple of 64 actually helps improve the efficiency. Reference: https://twitter.com/karpathy/status/1621578354024677377?s=20

Rocketknight1 replied:
I think we can just use a multiple of 64 anyway, it's not really a big change! The next multiple of 64 after 10000 is 10048.

sayakpaul replied:
Do we want to me to me to train the tokenizer and redo the TFRecords with that?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21657). All of your documentation changes will be reflected on that endpoint.

sayakpaul committed:
add: script for preparing the TFRecord shards.

sayakpaul committed:
add: sequence of execution to readme.

sayakpaul commented on examples/tensorflow/tpu/language-modeling/prepare_tfrecord_shards.py:
```diff
@@ -0,0 +1,142 @@
+        default="sayakpaul/unigram-tokenizer-wikitext",
+        help="Tokenizer identifier. Can be a local filepath or a Hub identifier.",
+    )
+    parser.add_argument(
+        "--shard_size",
+        type=int,
+        default=1000,
+        help="Number of entries to go in a single shard.",
+    )
```
We should likely follow some advice from [this guide](https://www.tensorflow.org/tutorials/load_data/tfrecord) to decide this number when running things at the full scale.

sayakpaul committed:
remove limit from the tfrecord shard name.

Rocketknight1 committed:
Add initial train_model.py

Rocketknight1 committed:
Add basic training arguments and model init

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,130 @@
+    tpu = initialize_tpu(args)
+    strategy = tf.distribute.experimental.TPUStrategy(tpu)
+
+    if args.bfloat16:
+        tf.keras.mixed_precision.set_global_policy("mixed_bfloat16")
```
Is it explicitly needed?

Rocketknight1 replied:
I believe TPU training takes place in float32 if that argument isn't specified!

Rocketknight1 committed:
Get up to the point of writing the data collator

Rocketknight1 committed:
Pushing progress so far!

Rocketknight1 committed:
Complete first draft of model training code

sayakpaul committed:
feat: grouping of texts efficiently.

Co-authored-by: Matt <rocketknight1@gmail.com>

sayakpaul commented:
@Rocketknight1 I incorporated the `group_texts()` utility that we discussed over Slack. Let me know if the changes look good to you. Most of it is copy-pasted from [here](https://github.com/huggingface/notebooks/blob/main/examples/language_modeling-tf.ipynb). 

[Here's](https://colab.research.google.com/gist/sayakpaul/adfaa9b45c9b56f222487995d0971645/scratchpad.ipynb) Colab Notebook where I verified these.

Rocketknight1 committed:
Add proper masking collator and get training loop working

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,278 @@
+        optimizer, schedule = create_optimizer(
+            num_train_steps=total_train_steps,
+            num_warmup_steps=total_train_steps // 20,
+            init_lr=args.learning_rate,
```
Maybe also scale the learning rate with batch size?

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,278 @@
+        raise ValueError(f"No .tfrecord files found in {args.train_dataset}.")
+    training_records = tf.data.Dataset.from_tensor_slices(training_records).shuffle(len(training_records))
+    train_dataset = tf.data.TFRecordDataset(training_records, num_parallel_reads=tf.data.experimental.AUTOTUNE)
+    train_dataset = train_dataset.shuffle(args.shuffle_buffer_size)
```
Why another shuffle?

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,278 @@
+    if not training_records:
+        raise ValueError(f"No .tfrecord files found in {args.train_dataset}.")
+    training_records = tf.data.Dataset.from_tensor_slices(training_records).shuffle(len(training_records))
+    train_dataset = tf.data.TFRecordDataset(training_records, num_parallel_reads=tf.data.experimental.AUTOTUNE)
```
We can shorten `tf.data.experimental.AUTOTUNE` to `tf.data.AUTOTUNE`. We could also consider using [`interleave()`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#interleave) while reading the TFRecords.

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,278 @@
+    # Many of the data collators in Transformers are TF-compilable when return_tensors == "tf", so we can
+    # use their methods in our data pipeline.
+    data_collator = DataCollatorForLanguageModeling(
+        tokenizer=tokenizer, mlm_probability=0.15, mlm=True, return_tensors="tf"
```
`mlm_probability` could be made a CLI argument.

sayakpaul commented:
@Rocketknight1 I took a deeper look into the TFRecord preparation script. I don't understand why there's a discrepancy in the following. 

While serializing the TFRecords, I am making each TFRecord shard has got a specific number of samples. When there are lesser samples for a TFRecord shard than the specified amount, that's fine. 

But when I load the TFRecords back and create a `tf.data.Dataset` out of them, the number of entries in the dataset (before batching) is much lesser. 


Here is a minimal Colab Notebook that demonstrates the issue: https://colab.research.google.com/gist/sayakpaul/b4b02f3f656c0041c93f6ba78c8e65fd/scratchpad.ipynb.

When you get a moment, could you take a look?

sayakpaul commented:
Thanks @Rocketknight1 for your help in debugging https://github.com/huggingface/transformers/pull/21657#issuecomment-1468086926 (discussed internally via Slack). I am currently regenerating the TFRecord shards. I will update here once that's done.

sayakpaul committed:
fix: things.

sayakpaul commented:
@Rocketknight1 corrected TFRecord shards have been pushed to `gs://tf-tpu-training-resources`.

Here are the record counts per split:

* Train: 300917
* Validation: 626
* Test: 722

The TFRecords were generated with a block size of 512.

Rocketknight1 committed:
Read sample counts from filenames

Rocketknight1 committed:
Read sample counts from filenames

Rocketknight1 committed:
Draft README

Rocketknight1 commented on examples/tensorflow/tpu/language-modeling/train_unigram.py:
```diff
@@ -0,0 +1,116 @@
+    parser.add_argument(
+        "--vocab_size",
+        type=int,
+        default=10000,
```
```suggestion
        default=10048,
```

Rocketknight1 commented on examples/tensorflow/tpu/language-modeling/prepare_tfrecord_shards.py:
```diff
@@ -0,0 +1,176 @@
+    return fn
+
+
+def group_texts(block_size=128):
```
```suggestion
def group_texts(block_size):
```
Maybe remove the default block size here since it should always be specified?

Rocketknight1 reviewed (approved):
Left some comments, but overall the tokenizer / data preparation code looks good!

Rocketknight1 committed:
Improve TPU warning

Rocketknight1 committed:
Use distribute instead of distribute.experimental

sayakpaul commented on examples/tensorflow/tpu/language-modeling/train_model.py:
```diff
@@ -0,0 +1,282 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Script for preparing TFRecord shards for pre-tokenized examples."""
```
To be changed.

sayakpaul committed:
Apply suggestions from code review

Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>

sayakpaul commented:
@Rocketknight1 the training code looks good to me, except for a few things:

* Maybe we should scale the LR with the batch size? 
* Take `mlm_probability` as a CLI arg? 
* Modularize the dataset preparation code a bit? 

But all these are non-blockers. Let's do 4 - 5 training runs varying the number of epochs and the learning rate.

Rocketknight1 committed:
Modularize loading and add MLM probability as arg

Rocketknight1 commented:
@sayakpaul MLM probability added as an arg and I modularized the loading!

Rocketknight1 committed:
Merge remote-tracking branch 'origin/examples/tf-tpu' into examples/tf-tpu

sayakpaul commented:
@Rocketknight1 started a training run with:

```python
python3 train_model.py \
	--tokenizer tf-tpu/unigram-tokenizer-wikitext \
	--per_replica_batch_size 256 \
	--tpu_name local --tpu_zone us-central1 --gcp_project huggingface-ml --bfloat16 \
	--train_dataset gs://tf-tpu-training-resources/train --eval_dataset gs://tf-tpu-training-resources/validation \
	--num_epochs 100 \
	--output_dir roberta-base-epochs-100 --hub_model_id tf-tpu/roberta-base-epochs-100
```

sayakpaul committed:
minor refactoring to better use the cli args.

sayakpaul committed:
Merge branch 'main' into examples/tf-tpu

sayakpaul commented:
@Rocketknight1 here's the final model trained with the command from [here](https://github.com/huggingface/transformers/pull/21657#issuecomment-1483729424):

https://huggingface.co/tf-tpu/roberta-base-epochs-100

When you try out examples in the widget of the model page ^, pass `[MASK]` instead of the default `<mask`. The results are far from perfect (visible from the validation accuracy), though.

sayakpaul committed:
readme fillup.

sayakpaul committed:
include tpu and inference sections in the readme.

sayakpaul committed:
table of contents.

sayakpaul committed:
parallelize maps.

sayakpaul committed:
polish readme.

sayakpaul committed:
change script name to run_mlm.py

sayakpaul commented:
@Rocketknight1 could you review [this PR](https://huggingface.co/tf-tpu/roberta-base-epochs-500-no-wd/discussions/1)?

sgugger commented on examples/tensorflow/tpu/language-modeling/prepare_tfrecord_shards.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def main(args):
+    wikitext = datasets.load_dataset("wikitext", "wikitext-103-raw-v1", split=args.split)
```
Would be cool to allow any dataset here, not just wikitext.

sgugger reviewed (commented):
Thanks a lot for working on this! I left a couple of comments.

sayakpaul committed:
address PR feedback (round I).

sayakpaul commented:
@sgugger thanks!

I addressed your comments. For https://github.com/huggingface/transformers/pull/21657#discussion_r1164017322, I will defer to @Rocketknight1.

Rocketknight1 commented on examples/tensorflow/tpu/language-modeling/run_mlm.py:
```diff
@@ -0,0 +1,307 @@
+            ~tf.cast(batch["attention_mask"], tf.bool)
+            | (batch["input_ids"] == tokenizer.cls_token_id)
+            | (batch["input_ids"] == tokenizer.sep_token_id)
+        )
```
Hm, on second thoughts, fixing it would require regenerating and reuploading the whole dataset and then updating the training loop too. Think it's worth it?

sgugger replied:
Not necessarily, but it would be cleaner if you ever do a v2.

Rocketknight1 replied:
Noted, will do!

gante reviewed (approved):
🔥

sayakpaul commented:
Merging since the failing tests are unrelated.
